switchboard/sql/SCHEMAv2.sql

```diff
@@ -0,0 +1,249 @@
+-- -- Optional host output
+    host_output                 text,
+    terminated_at               timestamp with time zone,
+
+-- For bookkeeping purposes; for ease of use, for now this works by just putting `last_updated_at = DEFAULT` on UPDATE queries, as per
+-- https://www.morling.dev/blog/last-updated-columns-with-postgres/. The original intention was to have something like an ON UPDATE trigger,
+-- but since it seems a bit messy to do with Postgres, it's being left alone for now.
+    last_updated_at             timestamp with time zone         not null default current_timestamp,
+
+---->> INVARIANT CHECKING <<----
+
+-- There are two allowed states:
+--  (1) resume_job_id = null, image_id != null, restart_job_id = _
+--  (2) resume_job_id != null, image_id = null, restart_job_id = null
+    constraint valid_init_spec check
+        ((resume_job_id is null and (image_id is not null))
+            or (resume_job_id is not null and restart_job_id is null and image_id is null)),
+-- ImageId is 32 bytes
+    constraint valid_image_id check
+        (case
+             when image_id is not null then octet_length(image_id) = 32
+             else true
+            end),
+-- queued => started_at, R.O.S.I = null
+-- running => started_at, R.O.S.I != null
+    constraint valid_queued_implication check (functional_state != 'queued' or
+                                               (started_at is null and dispatched_on_supervisor_id is null)),
+    constraint valid_dispatched_implication check (functional_state != 'dispatched' or
+                                                   (started_at is not null and dispatched_on_supervisor_id is not null)),
+-- Restart count >= 0
+    constraint valid_restart_policy check ((restart_policy).remaining_restart_count >= 0),
+-- Exit status vs. host output
+    constraint exit_status_allows_host_output check (case
+                                                         when exit_status = 'supervisor_match_error'
+                                                             then host_output is null
+                                                         when exit_status = 'queue_timeout' then host_output is null
+                                                         when exit_status = 'job_timeout' then host_output is null
+                                                         when exit_status = 'job_canceled' then host_output is null
+                                                         when exit_status = 'supervisor_dropped_job'
+                                                             then host_output is null
+                                                         when exit_status is null then host_output is null
+                                                         else true
+        end),
+-- exit status != null <=> functional state == finalized
+    constraint exit_status_nullity_iso_finalized check ((exit_status is not null) = (functional_state = 'finalized'))
```

Comment:
This seems fine -- it implies that we'll only commit the exit status to the DB once the job has terminated.

An alternative design would simply always write the exit status to the DB (even a preliminary one), and simply refuse to update once finalized. With this current design, we'll have to cache the exit status in the supervisor's view of the job state and potentially (on reconnect) merge it with the switchboard's view. I'm not sure which one's better. Let's leave it at this for now.

Reply:
For this in particular, tentative status are stored in the job_events table, which is also used to determine the exit status in the jobs table during finalization.

Reply:
That makes a bunch of sense!